### ISSUE TITLE
New module: Split log softmax with loss

### ISSUE THREAD

**Issue by alisafaya** [pytorch/pytorch#61452]
## 🚀 Feature: SplitLogSoftmaxWithLoss

An approximation of Log Softmax based on [splitsoftmax](https://gist.github.com/alisafaya/785e431539917cfbaab23281b77699d9). 

## Motivation

The current and only existing softmax approximation module in pytorch [AdaptiveLogSoftmaxWithLoss](https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveLogSoftmaxWithLoss.html#torch.nn.AdaptiveLogSoftmaxWithLoss) implementation is that it has two downsides: 

- You cannot share weights with Embedding layer when using it for sequence modeling tasks.
- You are bound to pass your input through two sequential linear layers, even if you choose `div_value = 1.0` 

## Pitch

I would like to introduce `SplitLogSoftmaxWithLoss` module to PyTorch. This module calculates an approximation of `nn.CrossEntropyLoss`. This module splits the possible outputs of Softmax into multiple clusters the same way [AdaptiveLogSoftmaxWithLoss](https://pytorch.org/docs/stable/generated/torch.nn.AdaptiveLogSoftmaxWithLoss.html#torch.nn.AdaptiveLogSoftmaxWithLoss) does, the main difference is that `SplitLogSoftmaxWithLoss` does not use smaller sequential linear layers for projecting the input of Softmax into the final output, but it uses a single linear layer with the same size. This is similar to setting `div_value = 1.0` in `AdaptiveLogSoftmaxWithLoss`. But this way you can make use of sharing weights between the `Embedding` layer and the final `Projection` layer inside  `SplitLogSoftmaxWithLoss`. 

I have prepared a working implementation of it in this gist [splitsoftmax.py](https://gist.github.com/alisafaya/785e431539917cfbaab23281b77699d9) based on [splitcross](https://github.com/ari-holtzman/genlm/blob/master/splitcross.py). The original splitcross implementation doesn't support multiple splits `cutoffs`, where as the one I am sharing supports them.

**Comment by jbschlosser** [comment]
Hey @alisafaya, thanks for the request! Can you point out any papers that reference this loss function? In general, our policy is to accept new modules into PyTorch only once the value of doing so has been well established. We can certainly consider adding it if the approach is widely used or if this becomes a popular request.

**Comment by alisafaya** [comment]
Hi @jbschlosser, 

As far as I know, this loss is first proposed in this paper: https://arxiv.org/pdf/1803.08240.pdf (section 3.3). 

And it is used in other implementations like: 

- AWD LSTM: https://github.com/salesforce/awd-lstm-lm
- Single Headed Attention RNN https://arxiv.org/abs/1911.11423
